pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -662,6 +659,51 @@ public void run() {
       long catchUpTimeMillis = 0L;
       _startTimeMs = now();
       try {
+        // Check READY_TO_CONSUME_DATA before enter consumeLoop in the new PartitionConsumer.
+        if (!_isReadyToConsumeData.getAsBoolean()) {
+          do {
+            //noinspection BusyWait
+            Thread.sleep(RealtimeTableDataManager.READY_TO_CONSUME_DATA_CHECK_INTERVAL_MS);
+          } while (!_shouldStop && !endCriteriaReached() && !_isReadyToConsumeData.getAsBoolean());
+        }
+
+        if (_tableConfig.getUpsertConfig() != null && _tableConfig.getUpsertConfig().isEnableSnapshot()) {
+          // Persist snapshot for sealed segments. We need to guarantee the previous segment is already replaced with
+          // the immutable segment, so the snapshot might not be persisted for the previous consuming segment.
+          List<SegmentDataManager> allSegments = _realtimeTableDataManager.acquireAllSegments();
+          List<SegmentDataManager> allSegmentsForPartition = new ArrayList<>();
+          for (SegmentDataManager segmentDataManager: allSegments) {
+            // release segments not this partition
+            if (_partitionGroupId == new LLCSegmentName(segmentDataManager.getSegmentName()).getPartitionGroupId()) {
+              allSegmentsForPartition.add(segmentDataManager);
+            } else {
+              _realtimeTableDataManager.releaseSegment(segmentDataManager);
+            }
+          }
+          // wait if all segments (except the new consuming segment) for this partition not sealed completely.
+          // The only consuming segment should be the new consuming segment, all the other segments should be persisted.
+          if (allSegmentsForPartition.stream()
+              .filter(segmentDataManager -> segmentDataManager.getSegment().getSegmentMetadata().isMutableSegment())
+              .count() > 1) {
+            do {
+              //noinspection BusyWait
+              Thread.sleep(RealtimeTableDataManager.READY_TO_CONSUME_DATA_CHECK_INTERVAL_MS);
+            } while (allSegmentsForPartition.stream()
```

Comment:
(MAJOR) This will never return true without acquiring the segments again.
We should acquire the segments that are previous mutable in the loop, until there is no mutable segment.